src/main/java/part2/cache/CachingDataStorageImpl.java

```diff
@@ -38,6 +39,22 @@ public CachingDataStorageImpl(DataStorage<String, T> db, int timeout, TimeUnit t
         // TODO don't use obtrudeException on result - just don't
         // TODO use remove(Object key, Object value) to remove target value
         // TODO Start timeout after receiving result in CompletableFuture, not after receiving CompletableFuture itself
-        throw new UnsupportedOperationException();
+        return cache.computeIfAbsent(key, s -> {

```

Comment:
Don't use `computeIfAbsent` here. In case of `timeout == 0` you could get `cache.remove` call before the value placed into `Map`.